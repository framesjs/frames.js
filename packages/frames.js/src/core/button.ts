import type { FramePlainObjectButtonElement } from "./types";

/**
 * Defines a button in a typesafe way without the React requirement. This is just a helper to have IDE autocompletion.
 *
 * @example
 * ```ts
 * import { button } from "frames.js/core";
 *
 * {
 *    buttons: [
 *      button({
 *        action: "post",
 *        label: "Click me",
 *        target: "/",
 *    }),
 *   ],
 * }
 * ```
 */
export function button(
  props: FramePlainObjectButtonElement
): FramePlainObjectButtonElement {
  return props;
}
