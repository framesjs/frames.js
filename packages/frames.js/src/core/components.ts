import { JsonValue } from "./types";

type PostButtonProps = {
  /** A 256-byte string which is label of the button */
  children: string;
  action: "post";
  /**
   * Either full URL or relative path that will be resolved against current url and basePath
   * if omitted it will send use current url and path
   */
  target?: string;
  /**
   * State passed to the next frame, will be available on the context as pressedButton.state
   */
  state?: JsonValue;
};

type PostRedirectButton = {
  /** A 256-byte string which is label of the button */
  children: string;
  action: "post_redirect";
  /**
   * Either full URL or relative path that will be resolved against current url and basePath
   * if omitted it will send use current url and path
   */
  target?: string;
  /**
   * State passed to the next frame
   *
   * State passed to the next frame, will be available on the context as pressedButton.state
   */
  state?: JsonValue;
};

type MintButtonProps = {
  /** A 256-byte string which is label of the button */
  children: string;
  action: "mint";
  /** The target  property MUST be a valid [CAIP-10](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-10.md) address, plus an optional token_id . */
  target: string;
};

type LinkButtonProps = {
  /** A 256-byte string which is label of the button */
  children: string;
  action: "link";
  target: string;
};

type TxButtonProps = Pick<PostButtonProps, "state"> & {
  /** A 256-byte string which is label of the button */
  children: string;
  action: "tx";
  /**
   * URL which points to a valid Frame Transaction URL, which returns tx calldata.
   *
   * Either full URL or relative path that will be resolved against current url and basePath
   * if omitted it will send use current url and path
   */
  target: string;
  /**
   * URL where a frame message containing the transaction ID will be posted if the transaction succeeds.
   * Overrides the top level frame post_url.
   */
  post_url?: string;
};

export type ButtonProps =
  | PostButtonProps
  | PostRedirectButton
  | MintButtonProps
  | LinkButtonProps
  | TxButtonProps;

export const Button: React.FunctionComponent<ButtonProps> = () => {
  return null;
};
