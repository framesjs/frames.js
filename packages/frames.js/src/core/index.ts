export { createFrames, type DefaultMiddleware } from "./createFrames";
export { concurrentMiddleware } from "./middleware/concurrentMiddleware";
export { composeMiddleware } from "./composeMiddleware";
export { Button } from "./components";
export { redirect } from "./redirect";
export * as types from "./types";
export * from "./constants";
