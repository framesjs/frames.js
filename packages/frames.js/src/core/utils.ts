import { FrameDefinition, FrameRedirect, JsonValue } from "./types";

const buttonActionToCode = {
  post: "p",
  post_redirect: "pr",
};

const BUTTON_INFORMATION_SEARCH_PARAM_NAME = "__bi";
const BUTTON_STATE_SEARCH_PARAM_NAME = "__bs";

function isValidButtonIndex(index: any): index is 1 | 2 | 3 | 4 {
  return (
    typeof index === "number" &&
    !Number.isNaN(index) &&
    index >= 1 &&
    index <= 4
  );
}

function isValidButtonAction(action: any): action is "post" | "post_redirect" {
  return (
    typeof action === "string" &&
    (action === "post" || action === "post_redirect")
  );
}

/**
 * This function generates fully qualified URL for post and post_redirect buttons
 * that also encodes the button type and button value so we can use that on next frame.
 */
export function generatePostButtonTargetURL({
  buttonIndex,
  buttonAction,
  currentURL,
  target,
  basePath,
  state,
}: {
  buttonIndex: 1 | 2 | 3 | 4;
  buttonAction: "post" | "post_redirect";
  currentURL: URL;
  basePath: string;
  target: string | undefined;
  /**
   * Button state
   */
  state: JsonValue | undefined;
}): string {
  let url = new URL(currentURL);

  if (target) {
    // resolve target relatively to basePath
    const baseUrl = new URL(basePath, currentURL);
    const preformatted = baseUrl.pathname + "/" + target;
    const parts = preformatted.split("/").filter(Boolean);
    const finalPathname = parts.join("/");

    url = new URL(`/${finalPathname}`, currentURL);
  }

  // store what button has been clicked in the URL
  url.searchParams.set(
    BUTTON_INFORMATION_SEARCH_PARAM_NAME,
    `${buttonIndex}:${buttonActionToCode[buttonAction]}`
  );
  if (state !== undefined) {
    url.searchParams.set(BUTTON_STATE_SEARCH_PARAM_NAME, JSON.stringify(state));
  }

  return url.toString();
}

type ButtonInformation = {
  action: "post" | "post_redirect";
  index: 1 | 2 | 3 | 4;
  state?: JsonValue;
};

export function parseButtonInformationFromTargetURL(
  url: URL
): undefined | ButtonInformation {
  const buttonInformation = url.searchParams.get(
    BUTTON_INFORMATION_SEARCH_PARAM_NAME
  );
  const buttonState = url.searchParams.get(BUTTON_STATE_SEARCH_PARAM_NAME);

  // parse state only if button has been identified
  if (!buttonInformation) {
    return undefined;
  }

  const [buttonIndex, buttonActionCode] = buttonInformation.split(":");

  if (!buttonIndex || !buttonActionCode) {
    return undefined;
  }

  const index = parseInt(buttonIndex, 10);
  const action = Object.entries(buttonActionToCode).find(
    ([, value]) => value === buttonActionCode
  )?.[0] as "post" | "post_redirect";

  if (!isValidButtonIndex(index) || !isValidButtonAction(action)) {
    return undefined;
  }

  let state: JsonValue | undefined;

  if (buttonState) {
    try {
      state = JSON.parse(buttonState);
    } catch {
      console.warn("Failed to parse button state from URL");
    }
  }

  return {
    action,
    index,
    state,
  };
}

export function isFrameRedirect(value: any): value is FrameRedirect {
  return value && typeof value === "object" && value.kind === "redirect";
}

export function isFrameDefinition(value: any): value is FrameDefinition<any> {
  return value && typeof value === "object" && "image" in value;
}
