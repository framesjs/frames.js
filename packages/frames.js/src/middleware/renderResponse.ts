import { ImageResponse } from "@vercel/og";
import { getFrameHtmlHead } from "../getFrameHtml";
import { getFrameFlattened } from "../getFrameFlattened";
import { type Frame } from "../types";
import type { ButtonProps } from "../core/components";
import type {
  FrameButtonElement,
  FrameDefinition,
  FramesHandlerFunctionReturnType,
  FramesMiddleware,
} from "../core/types";
import {
  generatePostButtonTargetURL,
  generateTargetURL,
  isFrameRedirect,
} from "../core/utils";
import { FRAMES_META_TAGS_HEADER } from "../core/constants";
import { FrameMessageError } from "../core/errors";

class InvalidButtonShapeError extends Error {}

class InvalidButtonCountError extends Error {}

class UnrecognizedButtonActionError extends Error {}

class ImageRenderError extends Error {}

/**
 * This middleware is responsible for rendering the response
 *
 * If the accept header is set to application/json, it will return the metadata as JSON
 * so it is easy to parse it for metatags in existing applications.
 */
export function renderResponse(): FramesMiddleware<any, Record<string, any>> {
  return async (context, next) => {
    const wantsJSON =
      context.request.headers.get("accept") === FRAMES_META_TAGS_HEADER;
    let result: FramesHandlerFunctionReturnType<any> | Response | undefined;

    try {
      result = await next(context);
    } catch (e) {
      if (e instanceof FrameMessageError) {
        return Response.json(
          {
            message: e.message,
          },
          {
            status: e.status,
          }
        );
      }

      // eslint-disable-next-line no-console -- provide feedback to the user
      console.error(e);

      return new Response(
        wantsJSON
          ? JSON.stringify({ error: "Internal Server Error" })
          : "Internal Server Error",
        {
          status: 500,
          headers: {
            "Cache-Control": "no-store",
            "Content-Type": wantsJSON ? "application/json" : "text/plain",
          },
        }
      );
    }

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- this can happen if the handler returns undefined
    if (!result) {
      return new Response(
        wantsJSON
          ? JSON.stringify({ error: "Handler did not return a response" })
          : "Handler did not return a response",
        {
          status: 500,
          headers: {
            "Cache-Control": "no-store",
            "Content-Type": wantsJSON ? "application/json" : "text/plain",
          },
        }
      );
    }

    if (result instanceof Response) {
      return result;
    }

    if (isFrameRedirect(result)) {
      return new Response(null, {
        status: result.status ?? 302,
        headers: {
          ...result.headers,
          Location: result.location.toString(),
        },
      });
    }

    let state: string | undefined;

    // state is supported only in reactions to button clicks (POST requests)
    if (result.state) {
      if (context.request.method === "POST") {
        state = JSON.stringify(result.state);
      } else {
        // eslint-disable-next-line no-console -- provide feedback to the user
        console.warn(
          "State is not supported on initial request (the one initialized when Frames are rendered for first time, that uses GET request) and will be ignored"
        );
      }
    }

    try {
      // @todo validate frame so it is according to spec and throw a ValidationError in case it isn't
      // and handle that error in catch block
      const frame: Frame = {
        version: "vNext",
        state,
        // @todo rendering image could be moved to its own middleware instead so users can use something different if they want to?
        // but that would mean that we need to specify middleware manually in any app since importing it here in default middleware
        // and disabling it, has no effect on final bundle size of app
        image:
          typeof result.image === "string"
            ? generateTargetURL({
                target: result.image,
                baseUrl: context.baseUrl,
              }).toString()
            : await renderImage(result.image, result.imageOptions).catch(
                (e) => {
                  // eslint-disable-next-line no-console -- provide feedback to the user
                  console.error(e);

                  throw new ImageRenderError("Could not render image");
                }
              ),
        buttons: result.buttons
          ?.slice()
          .filter(
            (v): v is FrameButtonElement => typeof v === "object" && v !== null
          )
          .map((button, i): NonNullable<Frame["buttons"]>[number] => {
            if (!("type" in button && "props" in button)) {
              throw new InvalidButtonShapeError("Invalid button provided");
            }

            if (i > 3) {
              throw new InvalidButtonCountError("Up to 4 buttons are allowed");
            }

            const props = button.props as ButtonProps;

            switch (props.action) {
              case "link":
                return {
                  action: props.action,
                  label: props.children,
                  target: generateTargetURL({
                    target: props.target,
                    baseUrl: context.baseUrl,
                  }).toString(),
                };
              case "mint":
                return {
                  action: props.action,
                  label: props.children,
                  target: props.target,
                };
              case "tx":
                return {
                  action: props.action,
                  label: props.children,
                  target: generatePostButtonTargetURL({
                    buttonIndex: (i + 1) as 1 | 2 | 3 | 4,
                    buttonAction: "post",
                    target: props.target,
                    baseUrl: context.baseUrl,
                  }).toString(),
                  post_url: props.post_url
                    ? generatePostButtonTargetURL({
                        buttonIndex: (i + 1) as 1 | 2 | 3 | 4,
                        buttonAction: "post",
                        target: props.post_url,
                        baseUrl: context.baseUrl,
                      })
                    : undefined,
                };
              case "post":
              case "post_redirect":
                return {
                  action: props.action,
                  label: props.children,
                  target: generatePostButtonTargetURL({
                    buttonIndex: (i + 1) as 1 | 2 | 3 | 4,
                    buttonAction: props.action,
                    target: props.target,
                    baseUrl: context.baseUrl,
                  }),
                };
              default:
                throw new UnrecognizedButtonActionError(
                  "Unrecognized button action"
                );
            }
          }) as Frame["buttons"],
        inputText: result.textInput,
        imageAspectRatio: result.imageOptions?.aspectRatio ?? "1.91:1",
        accepts: result.accepts,
      };

      if (wantsJSON) {
        const flattened = getFrameFlattened(frame);

        return new Response(JSON.stringify(flattened), {
          status: result.status ?? 200,
          statusText: result.statusText ?? "OK",
          headers: {
            ...result.headers,
            "content-type": "application/json",
          },
        });
      }

      return new Response(getFrameHtmlHead(frame), {
        status: result.status ?? 200,
        statusText: result.statusText ?? "OK",
        headers: {
          ...result.headers,
          "content-type": "text/html",
        },
      });
    } catch (e) {
      let message = "Internal Server Error";

      // do not expose any unrecognized errors in response, use console.error instead
      if (
        e instanceof ImageRenderError ||
        e instanceof UnrecognizedButtonActionError ||
        e instanceof InvalidButtonCountError ||
        e instanceof InvalidButtonShapeError
      ) {
        message = e.message;
      } else {
        // eslint-disable-next-line no-console -- provide feedback to the user
        console.error(e);
      }

      return new Response(
        wantsJSON ? JSON.stringify({ error: message }) : message,
        {
          status: 500,
          headers: {
            "Cache-Control": "no-store",
            "Content-Type": wantsJSON ? "application/json" : "text/plain",
          },
        }
      );
    }
  };
}

async function renderImage(
  element: React.ReactElement,
  options: FrameDefinition<any>["imageOptions"]
): Promise<string> {
  const response = new ImageResponse(
    // do not use React jsx here because it causes problem with tree shaking React
    {
      type: "div",
      key: "",
      props: {
        style: {
          display: "flex", // Use flex layout
          flexDirection: "row", // Align items horizontally
          alignItems: "stretch", // Stretch items to fill the container height
          width: "100%",
          height: "100vh", // Full viewport height
          backgroundColor: "white",
        },
        children: [
          {
            type: "div",
            key: "",
            props: {
              style: {
                display: "flex",
                flexDirection: "column",
                justifyContent: "center",
                alignItems: "center",
                lineHeight: 1.2,
                fontSize: 36,
                color: "black",
                flex: 1,
                overflow: "hidden",
              },
              children: element,
            },
          },
        ],
      },
    },
    {
      ...(options?.aspectRatio === "1:1"
        ? {
            width: 1146,
            height: 1146,
          }
        : { width: 1146, height: 600 }),
      ...options,
    }
  );

  const buffer = await response.arrayBuffer();

  return `data:image/png;base64,${Buffer.from(buffer).toString("base64")}`;
}
