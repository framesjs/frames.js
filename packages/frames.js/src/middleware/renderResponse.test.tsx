/* eslint-disable no-console -- we are expecting console usage */
/* eslint-disable @typescript-eslint/require-await -- middleware expects async functions */
/* eslint-disable testing-library/render-result-naming-convention -- we are not using react testing library here */
import * as vercelOg from "@vercel/og";
import { FRAMES_META_TAGS_HEADER } from "../core";
import { Button } from "../core/components";
import { error } from "../core/error";
import { redirect } from "../core/redirect";
import type { FramesContext } from "../core/types";
import { resolveBaseUrl } from "../core/utils";
import { renderResponse } from "./renderResponse";

jest.mock("@vercel/og", () => {
  const arrayBufferMock = jest.fn(() => new ArrayBuffer(10));
  const constructorMock = jest.fn(
    () =>
      new (class {
        arrayBuffer = arrayBufferMock;
      })()
  );

  return {
    arrayBufferMock,
    constructorMock,
    ImageResponse: class {
      constructor(...args: unknown[]) {
        // @ts-expect-error -- we are mocking the constructor
        constructorMock(...args);
      }
      arrayBuffer = arrayBufferMock;
    },
  };
});

jest.mock(
  "../../package.json",
  () => ({
    version: "0.0.0-mock",
  }),
  { virtual: true }
);

describe("renderResponse middleware", () => {
  const arrayBufferMock: jest.Mock = (
    vercelOg as unknown as { arrayBufferMock: jest.Mock }
  ).arrayBufferMock;
  const constructorMock: jest.Mock = (
    vercelOg as unknown as { constructorMock: jest.Mock }
  ).constructorMock;
  const render = renderResponse();
  const request = new Request("https://example.com");
  const context: FramesContext<undefined> = {
    basePath: "/",
    initialState: undefined,
    request,
    url: new URL("https://example.com"),
    baseUrl: resolveBaseUrl(request, undefined, "/"),
  };

  beforeEach(() => {
    arrayBufferMock.mockClear();
    constructorMock.mockClear();
    context.basePath = "/";
    context.request = new Request("https://example.com");
    context.url = new URL("https://example.com");
    context.baseUrl = resolveBaseUrl(context.request, undefined, "/");
  });

  it("returns redirect Response if redirect is returned from handler", async () => {
    const result = await render(context, async () => {
      return redirect("https://example.com");
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(302);
    expect((result as Response).headers.get("location")).toBe(
      "https://example.com"
    );
  });

  it("allows custom status code and headers for redirect", async () => {
    const result = await render(context, async () => {
      return redirect("https://example.com", {
        status: 307,
        headers: { "X-Test": "test" },
      });
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(307);
    expect((result as Response).headers.get("location")).toBe(
      "https://example.com"
    );
    expect((result as Response).headers.get("x-test")).toBe("test");
  });

  it("allows to render a Response frame with custom status code and headers", async () => {
    const result = await render(context, async () => {
      return new Response("Hello", {
        status: 201,
        headers: { "X-Test": "test" },
      });
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(201);
    expect((result as Response).headers.get("x-test")).toBe("test");
  });

  it("allows to render a frame with image defined as react element", async () => {
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          <Button
            action="post"
            key="1"
            target={{ query: { value: "customStateValue1" } }}
          >
            Click me 1
          </Button>,
          <Button
            action="post"
            key="2"
            target={{ query: { a: true }, pathname: "/a/b" }}
          >
            Click me 2
          </Button>,
          <Button
            action="post"
            key="3"
            target={{ pathname: "/test", query: { value: 10 } }}
          >
            Click me 3
          </Button>,
        ],
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);
    await expect((result as Response).text()).resolves.toMatchSnapshot();
  });

  it("allows to render a frame as json if Accept header is used", async () => {
    context.request = new Request("https://example.com", {
      headers: {
        Accept: FRAMES_META_TAGS_HEADER,
      },
    });
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          <Button
            action="post"
            key="1"
            target={{ query: { value: "customStateValue1" } }}
          >
            Click me 1
          </Button>,
        ],
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);
    expect((result as Response).headers.get("Content-Type")).toBe(
      "application/json"
    );
    await expect((result as Response).text()).resolves.toMatchSnapshot();
  });

  it("properly resolves against baseUrl", async () => {
    const newContext = {
      ...context,
      baseUrl: new URL("https://example.com/prefixed"),
    };
    const result = await render(newContext, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          <Button
            action="post"
            key="1"
            target={{
              pathname: "/test",
              query: { value: "customStateValue1" },
            }}
          >
            Click me 1
          </Button>,
        ],
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);
    await expect((result as Response).text()).resolves.toMatchSnapshot();
  });

  it("renders text input", async () => {
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        textInput: "My input",
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);
    await expect((result as Response).text()).resolves.toMatchSnapshot();
  });

  it("returns error 500 if image rendering fails", async () => {
    arrayBufferMock.mockRejectedValueOnce(
      new Error("Something failed during render")
    );
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(500);
    await expect((result as Response).text()).resolves.toBe(
      "Could not render image"
    );
  });

  it("returns application error if error function is called", async () => {
    const result = await render(context, async () => {
      error("Custom error message");
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(400);
    await expect((result as Response).json()).resolves.toEqual({
      message: "Custom error message",
    });
  });

  it("returns application error if error function is called with custom status code", async () => {
    const result = await render(context, async () => {
      error("Custom error message", 401);
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(401);
    await expect((result as Response).json()).resolves.toEqual({
      message: "Custom error message",
    });
  });

  it("does not allow application errors with status codes other than 4XX", async () => {
    const result1 = await render(context, async () => {
      // @ts-expect-error -- we are providing invalid status code
      error("Custom error message", 200);
    });

    expect(result1).toBeInstanceOf(Response);
    expect((result1 as Response).headers.get("Content-Type")).toBe(
      "text/plain"
    );
    expect((result1 as Response).status).toBe(500);
    await expect((result1 as Response).text()).resolves.toBe(
      "Internal Server Error"
    );

    const result2 = await render(context, async () => {
      // @ts-expect-error -- we are providing invalid status code
      error("Custom error message", 500);
    });

    expect(result2).toBeInstanceOf(Response);
    expect((result2 as Response).headers.get("Content-Type")).toBe(
      "text/plain"
    );
    expect((result2 as Response).status).toBe(500);
    await expect((result2 as Response).text()).resolves.toBe(
      "Internal Server Error"
    );
  });

  it("returns 500 if invalid number of buttons is provided", async () => {
    // @ts-expect-error -- we are providing more than 4 buttons
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          <Button action="post" key="1">
            Click me 1
          </Button>,
          <Button action="post" key="2">
            Click me 2
          </Button>,
          <Button action="post" key="3">
            Click me 3
          </Button>,
          <Button action="post" key="4">
            Click me 4
          </Button>,
          <Button action="post" key="5">
            Click me 5
          </Button>,
        ],
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(500);
    expect((result as Response).headers.get("Content-Type")).toBe("text/plain");
    await expect((result as Response).text()).resolves.toBe(
      "Up to 4 buttons are allowed"
    );
  });

  it("returns 500 if unrecognized button action is provided", async () => {
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          // @ts-expect-error -- props are not matching the expected type
          <Button action="invalid" key="1">
            Click me 1
          </Button>,
        ],
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(500);
    await expect((result as Response).text()).resolves.toBe(
      "Unrecognized button action"
    );
  });

  it("returns 500 if invalid button shape is provided", async () => {
    // @ts-expect-error -- returns invalid object shape
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          // this is not a proper object created by React.createElement()
          { action: "link", props: "invalid" },
        ],
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(500);
    await expect((result as Response).text()).resolves.toBe(
      "Invalid button provided"
    );
  });

  it("correctly renders image wich conditional content", async () => {
    const newContext = context;

    let result = await render(newContext, async () => {
      return {
        image: (
          <span>
            {"something" in newContext
              ? "Something is there"
              : "Something is not there"}
          </span>
        ),
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);

    expect(constructorMock).toHaveBeenCalledTimes(1);
    expect(constructorMock).toHaveBeenCalledWith(
      expect.any(Object),
      expect.any(Object)
    );
    expect((constructorMock.mock.calls[0] as unknown[])[0]).toMatchSnapshot();

    // @ts-expect-error -- we are modifying the context object
    newContext.something = "true";

    result = await render(newContext, async () => {
      return {
        image: (
          <span>
            {"something" in newContext
              ? "Something is there"
              : "Something is not there"}
          </span>
        ),
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);

    expect(constructorMock).toHaveBeenCalledTimes(2);
    expect(constructorMock).toHaveBeenCalledWith(
      expect.any(Object),
      expect.any(Object)
    );
    expect((constructorMock.mock.calls[1] as unknown[])[0]).toMatchSnapshot();
  });

  it("correctly renders tx button", async () => {
    context.request = new Request("https://example.com", {
      headers: {
        Accept: FRAMES_META_TAGS_HEADER,
      },
    });
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          <Button action="tx" key="1" target="/tx" post_url="/txid">
            Tx button
          </Button>,
        ],
      };
    });

    const json = (await (result as Response).json()) as Record<string, string>;

    expect(json["fc:frame:button:1"]).toBe("Tx button");
    expect(json["fc:frame:button:1:action"]).toBe("tx");
    expect(json["fc:frame:button:1:target"]).toBe(
      "https://example.com/tx?__bi=1-p"
    );
    expect(json["fc:frame:button:1:post_url"]).toBe(
      "https://example.com/txid?__bi=1-p"
    );
  });

  it("returns 500 if state returned from next middleware is not a string", async () => {
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        state: { test: true },
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(500);
    await expect((result as Response).text()).resolves.toBe(
      "State must be a string"
    );
  });

  it("returns a state as is from next middleware", async () => {
    context.request = new Request("https://example.com", {
      method: "POST",
      headers: {
        Accept: FRAMES_META_TAGS_HEADER,
      },
    });
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        state: JSON.stringify({ test: true }),
      };
    });

    const json = (await (result as Response).json()) as Record<string, string>;

    expect(json["fc:frame:state"]).toEqual(JSON.stringify({ test: true }));
  });

  it("properly handles error from next middleware and returns response 500", async () => {
    const result = await render(context, async () => {
      throw new Error("Something went wrong");
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).headers.get("Content-Type")).toBe("text/plain");
    expect((result as Response).status).toBe(500);
    await expect((result as Response).text()).resolves.toBe(
      "Internal Server Error"
    );
  });

  it("properly handles error from next middleware and returns response 500 and JSON", async () => {
    context.request = new Request("https://example.com", {
      method: "POST",
      headers: {
        Accept: FRAMES_META_TAGS_HEADER,
      },
    });
    const result = await render(context, async () => {
      throw new Error("Something went wrong");
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).headers.get("Content-Type")).toBe(
      "application/json"
    );
    expect((result as Response).status).toBe(500);
    await expect((result as Response).text()).resolves.toBe(
      JSON.stringify({ error: "Internal Server Error" })
    );
  });

  it("properly renders button with target object containing query", async () => {
    const url = new URL("https://example.com");
    url.searchParams.set("some_existing_param", "1");

    context.request = new Request(url, {
      method: "POST",
      headers: {
        Accept: FRAMES_META_TAGS_HEADER,
      },
    });
    context.url = url;
    context.baseUrl = new URL("https://example.com/test");
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          <Button
            action="post"
            key="1"
            target={{ pathname: "/", query: { value: true } }}
          >
            Click me 1
          </Button>,
        ],
      };
    });

    const expectedUrl = new URL("/test", "https://example.com");
    expectedUrl.searchParams.append("value", "true");
    expectedUrl.searchParams.append("__bi", "1-p");

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);
    const json = (await (result as Response).json()) as Record<string, string>;
    expect(json["fc:frame:button:1:target"]).toBe(expectedUrl.toString());
  });

  it("ignores null/undefined buttons are correctly assigns indexes", async () => {
    context.request = new Request("http://example.com", {
      method: "POST",
      headers: {
        Accept: FRAMES_META_TAGS_HEADER,
      },
    });
    const result = await render(context, async () => {
      return {
        image: <div>My image</div>,
        buttons: [
          <Button
            action="post"
            key="1"
            target={{ pathname: "/", query: { value: true } }}
          >
            Click me 1
          </Button>,
          null,
          true,
          <Button action="post" key="2" target="/">
            Click me 2
          </Button>,
        ],
      };
    });

    expect(result).toBeInstanceOf(Response);
    expect((result as Response).status).toBe(200);
    const json = (await (result as Response).json()) as Record<string, string>;
    expect(json).toMatchObject({
      "fc:frame:button:1": "Click me 1",
      "fc:frame:button:1:target": expect.any(String) as string,
      "fc:frame:button:2": "Click me 2",
      "fc:frame:button:2:target": expect.any(String) as string,
    });
  });
});
