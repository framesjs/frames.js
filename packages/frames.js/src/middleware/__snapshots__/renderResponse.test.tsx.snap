// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderResponse middleware allows to render a frame as json if Accept header is used 1`] = `"{"fc:frame":"vNext","fc:frame:image":"data:image/png;base64,AAAAAAAAAAAAAA==","og:image":"data:image/png;base64,AAAAAAAAAAAAAA==","fc:frame:image:aspect_ratio":"1.91:1","fc:frame:button:1":"Click me 1","fc:frame:button:1:action":"post","fc:frame:button:1:target":"https://example.com/?value=customStateValue1&__bi=1-p","frames.js:version":"0.0.0-mock"}"`;

exports[`renderResponse middleware allows to render a frame with image defined as react element 1`] = `"<meta name="fc:frame" content="vNext"/><meta name="fc:frame:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="og:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="fc:frame:image:aspect_ratio" content="1.91:1"/><meta name="fc:frame:button:1" content="Click me 1"/><meta name="fc:frame:button:1:action" content="post"/><meta name="fc:frame:button:1:target" content="https://example.com/?value=customStateValue1&__bi=1-p"/><meta name="fc:frame:button:2" content="Click me 2"/><meta name="fc:frame:button:2:action" content="post"/><meta name="fc:frame:button:2:target" content="https://example.com/a/b?a=true&__bi=2-p"/><meta name="fc:frame:button:3" content="Click me 3"/><meta name="fc:frame:button:3:action" content="post"/><meta name="fc:frame:button:3:target" content="https://example.com/test?value=10&__bi=3-p"/><meta name="frames.js:version" content="0.0.0-mock"/>"`;

exports[`renderResponse middleware correctly renders image wich conditional content 1`] = `
{
  "key": "",
  "props": {
    "children": [
      {
        "key": "",
        "props": {
          "children": <span>
            Something is not there
          </span>,
          "style": {
            "alignItems": "center",
            "color": "black",
            "display": "flex",
            "flex": 1,
            "flexDirection": "column",
            "fontSize": 36,
            "justifyContent": "center",
            "lineHeight": 1.2,
            "overflow": "hidden",
          },
        },
        "type": "div",
      },
    ],
    "style": {
      "alignItems": "stretch",
      "backgroundColor": "white",
      "display": "flex",
      "flexDirection": "row",
      "height": "100vh",
      "width": "100%",
    },
  },
  "type": "div",
}
`;

exports[`renderResponse middleware correctly renders image wich conditional content 2`] = `
{
  "key": "",
  "props": {
    "children": [
      {
        "key": "",
        "props": {
          "children": <span>
            Something is there
          </span>,
          "style": {
            "alignItems": "center",
            "color": "black",
            "display": "flex",
            "flex": 1,
            "flexDirection": "column",
            "fontSize": 36,
            "justifyContent": "center",
            "lineHeight": 1.2,
            "overflow": "hidden",
          },
        },
        "type": "div",
      },
    ],
    "style": {
      "alignItems": "stretch",
      "backgroundColor": "white",
      "display": "flex",
      "flexDirection": "row",
      "height": "100vh",
      "width": "100%",
    },
  },
  "type": "div",
}
`;

exports[`renderResponse middleware properly resolves against baseUrl 1`] = `"<meta name="fc:frame" content="vNext"/><meta name="fc:frame:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="og:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="fc:frame:image:aspect_ratio" content="1.91:1"/><meta name="fc:frame:button:1" content="Click me 1"/><meta name="fc:frame:button:1:action" content="post"/><meta name="fc:frame:button:1:target" content="https://example.com/prefixed/test?value=customStateValue1&__bi=1-p"/><meta name="frames.js:version" content="0.0.0-mock"/>"`;

exports[`renderResponse middleware renders text input 1`] = `"<meta name="fc:frame" content="vNext"/><meta name="fc:frame:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="og:image" content="data:image/png;base64,AAAAAAAAAAAAAA=="/><meta name="fc:frame:input:text" content="My input"/><meta name="fc:frame:image:aspect_ratio" content="1.91:1"/><meta name="frames.js:version" content="0.0.0-mock"/>"`;
